mignotte bound applied in `dup_zz_zassenhaus(f, K)`

lagamura opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Application of PR 19254

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #19253 partially.

#### Brief description of what is fixed or changed
function `dup_zz_zassenhaus(f, K)` instead of calling `dup_zz_maignotte_bound(f, K)` calculates locally the old bound : `(abs(K.sqrt(K(n + 1))*2**n*A*b).`

#### Other comments
By changing the aforementioned line, there is an improvement on `factor(f)` to 10% on time. 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
   * improvement of `dup_zz_zassenhaus(f, K)`
<!-- END RELEASE NOTES -->

lagamura committed:
Update factortools.py

lagamura committed:
Merge branch '19253_mignotte_bound' of https://github.com/lagamura/sympy into 19253_mignotte_bound

lagamura commented:
@jksuom give a look on this

jksuom commented on sympy/polys/factortools.py:
```diff
@@ -323,7 +323,7 @@ def dup_zz_zassenhaus(f, K):
     A = dup_max_norm(f, K)
     b = dup_LC(f, K)
-    B = int(abs(K.sqrt(K(n + 1))*2**n*A*b))
+    B = int(dup_zz_mignotte_bound(f, K))
```
It looks like the improved Mignotte bound cannot be used here. There is a test https://github.com/sympy/sympy/blob/0515aeec9ff25f5abb831f0d54a6eb6704084551/sympy/polys/factortools.py#L402
that fails if `B` is too small. I haven't seen the original documentation but it seems that this `B` should not be changed.

lagamura committed:
added abs on B

lagamura committed:
2*bound

lagamura committed:
alkis_recommendation

This should probably wont work

lagamura committed:
Merge pull request #1 from lagamura/alkis

alkis_recommendation

jksuom commented on sympy/polys/factortools.py:
```diff
@@ -323,7 +323,11 @@ def dup_zz_zassenhaus(f, K):
     A = dup_max_norm(f, K)
     b = dup_LC(f, K)
-    B = 2*int(abs(dup_zz_mignotte_bound(f, K)))
+    if Poly(f).is_univariate:
```
There is no need for this test. This method is for univariate polynomials.

lagamura committed:
Revert "Merge pull request #1 from lagamura/alkis"

This reverts commit 02ada070dc966c1eb066bb9b017675c12f920d92, reversing
changes made to acdca0fc45f7ba940acba865a3025f15f489a4c7.

lagamura committed:
test with 3*mignotte

lagamura committed:
import mpmath trial

lagamura committed:
Revert "import mpmath trial"

This reverts commit 5c225678d466f3dfa2facf334a7b3b29d6595412.

lagamura commented:
Can someone help with that log: (link)[https://travis-ci.org/github/sympy/sympy/jobs/686930526]
Type error seems strange as we have not change anything on types of the function.

lagamura committed:
Revert "test with 3*mignotte"

This reverts commit 7de4b1ddb238b74c1ad7f05ed3ac5aae5f661d42.

lagamura commented:
It seems we are failing in some specific examples on travic ci. Although when we are running them locally we are passing them successfully as you can see by the attached file. Is there any possibility that travic ci has a bug?
[test2_latest.zip](https://github.com/sympy/sympy/files/4627632/test2_latest.zip)

lagamura committed:
Increasing c* mignotte_bound

It is in great interest that by increasing c from 3 to 4 we are passing a "type" error test

lagamura committed:
u == 0 case on dmp_mignotte

lagamura committed:
5*mig

by increasing again from 4 to 5 we are passing an error

lagamura committed:
Update factortools.py

lagamura committed:
Merge branch '19253_mignotte_bound' of https://github.com/lagamura/sympy into 19253_mignotte_bound

lagamura committed:
unused l1 function removed

lagamura commented:
We made some changes, and now Travis CI gives timeout because no output was recieved.
This means something like an infinite loop? How can i debug this?

jksuom commented:
It looks like that is happening in polytools (and maybe elsewhere). I would run `bin/test -v polytools` in order to see which test is hanging. Then I would run that test in a debugger and  interrupt it after a while. The stack trace will usually show periodicity that will help to locate the looping lines. I this case, those will probably be the changed lines so that putting a breakpoint there could also help.

smichr commented:
Just read an [article that might be of interest](https://www.quantamagazine.org/new-math-measures-the-repulsive-force-within-polynomials-20200514/) in which in says,

>Dimitrov proved that all non-cyclotomic polynomials must have a root that’s at least log 2 divided by four times the degree of the polynomial outside the unit circle; that is, (log 2)/(4d)

lagamura committed:
Update factortools.py

jksuom commented on sympy/polys/factortools.py:
```diff
@@ -169,7 +169,7 @@ def dup_zz_mignotte_bound(f, K):
 
     # euclidean-norm
-    eucl_norm = K.sqrt( sum( [cf**2 for cf in f] ) )
+    eucl_norm = _sqrt( sum( [cf**2 for cf in f] ) )
```
This will change the type of `eucl_norm` that is an element of `K`. What is the purpose of the change?

lagamura replied:
Alkis asked me to change this, because he founds that now it is passing an test. Although for me it stacks again in an infinite loop on `test_Poly_lift` file. Now i will try to find the periodicity you mentioned above

czgdp1807 commented:
@lagamura Are you still working on this? Thanks for your contributions. Please do not close this PR.

jksuom commented:
I think that this can be closed. It was based on a bound that did not work in this connection.

jksuom commented:
Closing.
